### PR TITLE
Implement upload from stream for IAssetFile

### PR DESCRIPTION
**Pull request by luminescent**
This pull request adds functionality for uploading an asset file sourced from a stream, rather than a file. This scenario is needed in the company I work for since we source our assets from other blobs in the cloud. We also have a requirement of "encryption at rest" (coming from the studios that provide us with media content), so whatever gets uploaded into Azure Media Services must have StorageEncryption. Since the SDK does the encryption at rest automatically on upload, we'd rather use that than the REST API to push asset content. 

Regarding the changes you'll see in the pull request, there are some areas of "wall of pink", please excuse that (got tricked by my automatic source code formatting). I tried to fixed them in the second commit, where I reverted to the original formatting.
 
Some of the code is duplicated between the methods that upload from a file and the ones I added, which upload from a stream. I could fix that, if you consider this functionality is useful in the SDK and want to include it. 

I have added tests in the Scenario project for this new functionality. They accompany the scenarios played out for uploading a file. I have not tested the new functionality for very large files (say over 5GB). 

The functionality is exposed via the method `IAssetFile.Upload(string name, Stream stream)` and its asynchronous counterpart. The `name` parameter is required and it's what is being displayed in the Azure Portal as the name for this particular asset file. Also it is used to compute the `InitializationVector` for the storage encryption.

### CODE REVIEW

**Comment by johndeu** [comment]
We are reviewing your Pull request. Thank you for contributing to the SDK!

File: test/net/Scenario/AssetTests.cs [hunk 1]
```diff
@@ -115,6 +115,53 @@ public void ShouldCreateAssetFile()
...
+
+        /// <summary>
+        /// Known issue with mime-type detection
+        /// </summary>
+        [TestMethod]
+        [TestCategory("ClientSDK")]
+        [Owner("ClientSDK")]
+        [Priority(1)]
+        [DeploymentItem(@"Media\SmallWmv.wmv", "Media")]
+        public void ShouldCreateAssetFileFromStream()
```
Comment (quintinb) [inline test/net/Scenario/AssetTests.cs]: Can you add test cases for the zero byte stream and a stream with a file size larger than 32MB to cover all three cases in the upload code?

Comment (quintinb) [inline test/net/Scenario/AssetTests.cs]: Also, could you add a test case that uses a network stream instead of using all file streams?  Perhaps upload a file to blob storage and then download it.  Pass the download stream to the upload method.  Not a real world case but could simulate a flle uploaded to a web role that was then used to create an asset with a network stream.

File: test/net/Scenario/AssetTests.cs [hunk 2]
```diff
@@ -944,13 +1031,97 @@ private void VerifyAndDownloadAsset(IAsset asset, int expectedFileCount,string i
             }
         }
 
+        private void VerifyAndDownloadAsset(IAsset asset, int expectedFileCount, Stream inputStream, bool inputFileValidation, bool performStorageSdkDownloadVerification = true)
```
Comment (quintinb) [inline test/net/Scenario/AssetTests.cs]: Instead of duplicating so much code can you have the Stream overload be the primary and have another that takes a file and creates a filestream?  Or have the MD5 hash calculated by the caller and passed in?

File: test/net/Scenario/AssetTests.cs [hunk 3]
```diff
@@ -545,6 +612,26 @@ public void ShouldDownloadEnvelopeEncryptionProtectedAssetFile()
...
         }
 
+
+        [TestMethod]
+        [TestCategory("ClientSDK")]
+        [Owner("ClientSDK")]
+        [Priority(1)]
+        [DeploymentItem(@"Media\SmallWmv.wmv", "Media")]
+        [TestCategory("Bvt")]
+        public void ShouldDownloadEnvelopeEncryptionProtectedAssetFileCreatedFromStream()
```
Comment (quintinb) [inline test/net/Scenario/AssetTests.cs]: Instead of this method could you add a "FromStream" version of ShouldCreateEncryptedInitilizedAsset?  Or add asset validation to one of the other methods you added to ensure that storage encryption is working properly with Stream based uploads.

File: src/net/Client/Entities/IAssetFile.cs [hunk 4]
```diff
@@ -93,11 +95,16 @@ public partial interface IAssetFile
         /// <returns>A function delegate that returns the future result to be available through the Task.</returns>
         Task UploadAsync(string path, BlobTransferClient blobTransferClient, ILocator locator, CancellationToken token);
 
+        Task UploadAsync(string name, Stream stream, BlobTransferClient blobTransferClient, ILocator locator,
```
Comment (quintinb) [inline src/net/Client/Entities/IAssetFile.cs]: The comment on the locator parameter starts with "An asset ...".  I think you mean "A locator ..." instead.

File: src/net/Client/Entities/AssetFileData.cs [hunk 5]
```diff
@@ -291,9 +292,76 @@ public Task UploadAsync(string path, BlobTransferClient blobTransferClient, ILoc
                 });
         }
 
+       public Task UploadAsync(string name, Stream stream, BlobTransferClient blobTransferClient, ILocator locator, CancellationToken token)
```
Comment (quintinb) [inline src/net/Client/Entities/AssetFileData.cs]: The comment on the locator parameter starts with "An asset ...".  I think you mean "A locator ..." instead.

**Comment by luminescent** [comment]
Hello 

I have been through your suggestions and made the changes accordingly: removed the duplicated code, consolidated the logic into the stream based method, fixed the comments and added the extra tests, including the upload from a network stream. 

There are still a few points I am unsure of: 

- Please see here: https://github.com/luminescent/azure-sdk-for-media-services/blob/dev/src/net/Client/Common/Common.BlobTransfer/BlobUploader.cs#L221 . Should we keep the logic where we name the blob in Azure Media storage based on an extracted file name from whatever it's being passed in as the name, even when uploading from stream (where we're free to put in any string, which could be something other than a file name)? 
- Any best practices recommendations for cleaning up after tests? E.g. here: https://github.com/luminescent/azure-sdk-for-media-services/blob/dev/test/net/Scenario/AssetTests.cs#L313 . I'm uploading a file to a blob then using it for the network stream test; if the test passes, I'm deleting the uploaded blob, but if it does not, I don't because the exception from the failed Assert is thrown. Is there a consistent model you're using throughout the code base to handle such situations?
- The mime type for the asset file is inferred from the name passed into the Upload method (https://github.com/luminescent/azure-sdk-for-media-services/blob/dev/src/net/Client/Collections/AssetFileCollection.cs#L105). Since we can pass any name for the Upload from stream overload, are we happy to it being set to null when clients put in names that are not file names?

**Comment by quintinb** [comment]
Thanks for the updates, they look great.

The name of the blob should be IAssetFile.Name and that name should be usable as a filename since virtual machines in Azure Media Services may need to download the file and store it temporarily for processing or end customers may progressively download the file for playback.  I am going to remove the name parameter from the Upload and UploadAsync overloads and just use IAssetFile.Name.

The tests do not do as good of a job of cleaning up after themselves as they should in many places.  The correct thing to do is add try/finally logic and I can add that to the test you are calling out.

Setting the mimetype based on IAsset.Name is a valid assumption.  Note that the SDK user can always set the IAssetFile.MimeType explicitly and update the entity.

**Comment by quintinb** [comment]
This is now commited as https://github.com/Azure/azure-sdk-for-media-services/commit/f9179b0042404bfbbc67a876fc579f85203ca68f.
